crates/typst-library/src/text/shaping.rs

```diff
@@ -994,3 +981,76 @@ fn assert_glyph_ranges_in_order(glyphs: &[ShapedGlyph], dir: Dir) {
         }
     }
 }
+
+/// Whether the glyph is a space.
+#[inline]
+fn is_space(c: char) -> bool {
+    matches!(c, ' ' | '\u{00A0}' | '　')
+}
+
+/// Whether the glyph is part of a CJK script.
+#[inline]
+fn is_cjk_script(c: char) -> bool {
+    use Script::*;
+    // U+30FC: Katakana-Hiragana Prolonged Sound Mark
+    matches!(c.script(), Hiragana | Katakana | Han) || c == '\u{30FC}'
+}
+
+/// See <https://www.w3.org/TR/clreq/#punctuation_width_adjustment>
+#[inline]
+fn is_cjk_left_aligned_punctuation(
+    c: char,
+    x_advance: Em,
+    stretchability: (Em, Em),
+    gb_style: bool,
+) -> bool {
+    // CJK quotation marks shares codepoints with latin quotation marks.
+    // But only the CJK ones have full width.
+    if matches!(c, '”' | '’') && x_advance + stretchability.1 == Em::one() {
+        return true;
+    }
+
+    if gb_style && matches!(c, '，' | '。' | '、' | '：' | '；') {
+        return true;
+    }
+
+    matches!(c, '》' | '）' | '』' | '」')
+}
+
+/// See <https://www.w3.org/TR/clreq/#punctuation_width_adjustment>
+#[inline]
+fn is_cjk_right_aligned_punctuation(
+    c: char,
+    x_advance: Em,
+    stretchability: (Em, Em),
+) -> bool {
+    // CJK quotation marks shares codepoints with latin quotation marks.
+    // But only the CJK ones have full width.
+    if matches!(c, '“' | '‘') && x_advance + stretchability.0 == Em::one() {
+        return true;
+    }
+
+    matches!(c, '《' | '（' | '『' | '「')
+}
+
+/// See <https://www.w3.org/TR/clreq/#punctuation_width_adjustment>
+#[inline]
+fn is_cjk_center_aligned_punctuation(c: char, gb_style: bool) -> bool {
+    if !gb_style && matches!(c, '，' | '。' | '、' | '：' | '；') {
+        return true;
+    }
+
+    // U+30FB: Katakana Middle Dot
+    matches!(c, '\u{30FB}')
+}
+
+/// Whether the glyph is justifiable.
+#[inline]
+fn is_justifiable(c: char, x_advance: Em, stretchability: (Em, Em)) -> bool {
```

Comment:
I just realized that this cannot be blindly cached because x_advance and stretchablity can change when adjusting for cjk specific things...

Reply:
perhaps something like when they change, recompute `is_justifiable`

Reply:
Yes I know, I am fixing it ;) I am just updating the cache value when they change

Reply:
The cache should still take effect for all latin text and at least some CJK text I would assume